### PR TITLE
Docs: TAP filter

### PR DESCRIPTION
**Pull request by Nadine2016**
# Description

Please include a high level summary of the changes.

This bug fixes ... \ This new feature can be used to ...

_Fill out any of the following sections that are relevant and remove the others_
## API changes
- Added x field to y resource
- ...

## Code changes
- Fix error in `Foo()` function
- Add `Bar()` function
- ...

## CI changes
- Adjusted schedule for x job
- ...

## Docs changes
- Added guide about feature x to public docs
- Updated README to account for y behavior
- ...

# Context

Users ran into this bug doing ... \ Users needed this feature to ...

See slack conversation [here](https://solo-io-corp.slack.com/archives/some/post)

## Interesting decisions
 
We chose to do things this way because ...

## Testing steps

I manually verified behavior by ...

## Notes for reviewers

Be sure to verify intended behavior by ...

Please proofread comments on ...

This is a complex PR and may require a huddle to discuss ...

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: docs/content/guides/traffic_management/listener_configuration/tap.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
+---
+title: Traffic tapping
+weight: 100
+description: Copy contents of HTTP requests and responses to an external TAP server. 
+---
+
+Copy contents of HTTP or gRPC requests and responses to an external TAP server. A Test Access Point (TAP) server is a simple device that connects directly to your infrastructure and receives copies of actual traffic from your network so that this traffic can be further monitored, analyzed, or tested with.
```
Comment (ashishb-solo) [inline docs/content/guides/traffic_management/listener_configuration/tap.md:7]: Curious to know if you found this acronym in the Envoy docs, or some other authoritative source? To the best of my knowledge, it's just "tap" (not an acronym, but taken from the definition of tap as "extracting", or just wiretapping).

So I think you might have to change all instances of "TAP" to "tap", but would be interested to hear if you found a meaning of which I was not previously aware

File: docs/content/guides/traffic_management/listener_configuration/tap.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+       matchLabels:
+         app: sample-tap-server-http
+     replicas: 1
+     template:
+       metadata:
+         labels:
+           app: sample-tap-server-http
+       spec:
+         containers:
+         - image: gcr.io/solo-test-236622/sample-tap-server-http:0.0.2
```
Comment (artberger) [inline docs/content/guides/traffic_management/listener_configuration/tap.md:51]: Can we use the test project for prod docs? If not, maybe we could move to the `solo-public` project, such as the docs repo here: https://console.cloud.google.com/gcr/images/solo-public/global/docs?project=solo-public

File: docs/content/guides/traffic_management/listener_configuration/tap.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+           sinks:
+           - httpService:
+               tapServer:
+                 name: sample-tap-server-http
+                 namespace: gloo-system
+               timeout: '20s'
+   EOF
+   ```
+
+2. In a terminal window, tail the log of the TAP server.
```
Comment (artberger) [inline docs/content/guides/traffic_management/listener_configuration/tap.md:116]:
```suggestion
2. In a terminal window, tail the logs of the TAP server. 
```

File: docs/content/guides/traffic_management/listener_configuration/tap.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+Traffic tapping support in Gloo Edge is introduced as an **alpha feature**. Alpha features are likely to change, are not fully tested, and are not supported for production.
+{{% /notice %}}
+
+{{% notice note %}}
+Traffic tapping is an Enterprise-only feature. 
+{{% /notice %}}
+
+## About traffic tapping in Gloo Edge
+
+* Traffic tapping can be applied to a listener in a `Gateway` and `MatchableHTTPGateway` resource. As such, traffic tapping is applied to all routes that the gateway serves. You cannot set up traffic tapping for a specific route.
```
Comment (ashishb-solo) [inline docs/content/guides/traffic_management/listener_configuration/tap.md:19]: Could we remove the `MatchableHTTPGateway` support? It should work in theory, but it's not in our product roadmap right now, so I am hesitant to broadcast this as something that's officially supported

Comment (ashishb-solo) [inline docs/content/guides/traffic_management/listener_configuration/tap.md:19]: > You cannot set up traffic tapping for a specific route.

I would rephrase this as "Tapping traffic for a specific route is not currently supported" as this is something that we will probably implement as a part of the next release

File: docs/content/guides/traffic_management/listener_configuration/tap.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+{{% /notice %}}
+
+{{% notice note %}}
+Traffic tapping is an Enterprise-only feature. 
+{{% /notice %}}
+
+## About traffic tapping in Gloo Edge
+
+* Traffic tapping can be applied to a listener in a `Gateway` and `MatchableHTTPGateway` resource. As such, traffic tapping is applied to all routes that the gateway serves. You cannot set up traffic tapping for a specific route. 
+* You can send traffic copies to the TAP server by using the HTTP or gRPC protocol. Traffic copies are buffered in memory before they are sent to the TAP server.
```
Comment (ashishb-solo) [inline docs/content/guides/traffic_management/listener_configuration/tap.md:20]:
```suggestion
* Users are responsible for writing their own tap servers. The tap server definitions can be found in the [tap-extension-examples repository](https://github.com/solo-io/tap-extension-examples). In order to receive the tap traces, users will have to implement their own service to receive data over either gRPC or HTTP. In order to do so, the server must implement the [protobuf definitions provided here](https://github.com/solo-io/tap-extension-examples/tree/main/pkg/tap_service). **The `tap-extension-examples` repository is provided solely as a reference for users implementations and is not intended to be used in production.**
* In the current implementation, the data plane buffers all trace data in memory before sending it to the tap server.
* Data tapped from the data plane server may contain sensitive information including credentials or personally identifying information. **Users should take care to ensure that all data is encrypted in transit and that sensitive information is scrubbed by the tap server receiving the data before it is written to storage or relayed to an additional service.**
```

**Review by ashishb-solo — COMMENTED** [review]
Thanks Nadine! A few changes need to be made, but overall looks great. Ping me once you've addressed the comments that I posted and I'll take another look

File: docs/content/guides/traffic_management/listener_configuration/tap.md (hunk 1 above)
Comment (Nadine2016) [reply docs/content/guides/traffic_management/listener_configuration/tap.md:7]: Hey @ashishb-solo the Envoy docs do not say this, but if you search for traffic tapping, it will come up pretty often on network pages. Here are a few examples: 
- https://www.gigamon.com/resources/resource-library/white-paper/understanding-network-taps-first-step-to-visibility.html 
- https://www.garlandtechnology.com/network-tap
- https://www.techopedia.com/definition/25311/network-tap

But I am also happy to change that.

File: docs/content/guides/traffic_management/listener_configuration/tap.md (hunk 2 above)
Comment (ashishb-solo) [reply docs/content/guides/traffic_management/listener_configuration/tap.md:59]: since tap is in alpha for now, my inclination is to leave it as is and try to figure out a long term plan as a part of our scoping for the next release.

would like to hear what @nfuden thinks about this strategy

Comment (ashishb-solo) [reply docs/content/guides/traffic_management/listener_configuration/tap.md:59]: that said, i don't think it's big of a lift to do this now. my hesitation stems more from the fact that i'd rather have a long-term strategy first before making any changes, namely to avoid having to do multiple moves here

File: docs/content/guides/traffic_management/listener_configuration/tap.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+{{% /notice %}}
+
+{{% notice note %}}
+Traffic tapping is an Enterprise-only feature. 
+{{% /notice %}}
+
+## About traffic tapping in Gloo Edge
+
+* Traffic tapping can be applied to a listener in a `Gateway` resource. As such, traffic tapping is applied to all routes that the gateway serves. Tapping traffic for a specific route is not currently supported.
+* Users are responsible for writing their own tap servers. The tap server definitions can be found in the [`tap-extension-examples` repository](https://github.com/solo-io/tap-extension-examples). To receive tap traces, the tap server service must implement the [tap service protobuf definitions](https://github.com/solo-io/tap-extension-examples/tree/main/pkg/tap_service) and be configured to receive data over the gRPC or HTTP protocol.
```
Comment (ashishb-solo) [inline docs/content/guides/traffic_management/listener_configuration/tap.md:20]:
```suggestion
* Users are responsible for writing their own tap servers. The tap server definitions can be found in the [`tap-extension-examples` repository](https://github.com/solo-io/tap-extension-examples). To receive tap traces, the tap server must implement the [tap service protobuf definitions](https://github.com/solo-io/tap-extension-examples/tree/main/pkg/tap_service) and be configured to receive data over the gRPC or HTTP protocol.
```

File: docs/content/guides/traffic_management/listener_configuration/tap.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+  {{% notice note %}}
+  The `tap-extension-examples` repository is provided as an implementation reference. The repository is **not intended to be used in production**.
+  {{% /notice %}}
+* In the current implementation, the data plane buffers all trace data in memory before sending it to the tap server.
+* You cannot tap traffic to a local file. 
+
+{{% notice warning %}}
+Data that is tapped from the data plane might contain sensitive information, including credentials or Personal Identifiable Information (PII). Before using traffic tapping in your environment, make sure that all data is encrypted during transit and that sensitive data is masked or removed by the tap server before it is written to permanent storage or forwarded to another service. For example, you might use a [Data Loss Prevention (DLP) policy]({{< versioned_link_path fromRoot="/guides/security/data_loss_prevention/" >}}) to mask sensitive information.
```
Comment (ashishb-solo) [inline docs/content/guides/traffic_management/listener_configuration/tap.md:28]:
```suggestion
Data that is tapped from the data plane might contain sensitive information, including credentials or Personal Identifiable Information (PII). Before using traffic tapping in your environment, make sure that all data is encrypted during transit and that sensitive data is masked or removed by the tap server before it is written to permanent storage or forwarded to another service.
```

Actually, the DLP plugin would not scrub sensitive data from reaching the tap server, so we should remove that. If anything, I would include a disclaimer stating that the DLP plugin would not remove sensitive data from being leaked through the tap filter, something like this:

> Users should also be advised that the DLP plugin will not prevent sensitive data from being leaked via the tap filter

**Review by ashishb-solo — COMMENTED** [review]
Thanks Nadine! A few more changes.... could you ping me to review it again when you're done?

File: docs/content/guides/traffic_management/listener_configuration/tap.md (hunk 2 above)
Comment (Nadine2016) [reply docs/content/guides/traffic_management/listener_configuration/tap.md:59]: decision from product meeting on 12/15: we can publish this repo

**Review by artberger — APPROVED** [review]

**Review by Rachael-Graham — APPROVED** [review]
